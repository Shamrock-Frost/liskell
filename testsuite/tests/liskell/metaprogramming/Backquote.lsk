(defmodule Backquote _ (Liskell SrcLoc LskParseTree SimpleList))

(define (backquote-dispatcher-on-syms ks kf pt)
    (case pt
      ((PSym loc sym qual)
       (if (== (head sym) #\`)
	   (ks (PList loc ([] (PSym noSrcSpan "PSym" "")
				    (PSym noSrcSpan "noSrcSpan" "")
				    (PString noSrcSpan (tail sym))
				    (PString noSrcSpan qual))
			    noPP))
	   (kf pt)))
      (_ (kf pt))))


(defenv (lambda ((LskEnv e p t d)) 
	  (return (LskEnv (backquote-dispatcher-on-syms trf_expr e) p t d))))

(defenv (lambda ((LskEnv e p t d)) 
	  (return (LskEnv (simple-list-dispatcher trf_expr e) p t d))))

(define (sPList lst)
    (PList noSrcSpan lst noPP))

(define (bq-transform-lst pts curlst)
    (case pts
      (([]) (sPList (: `[] (reverse curlst))))
      ((: x xs)
       (case x
	 ;; Escaped symbol of the forms ,foo ,@foo
	 ((@ sym (PSym loc string qual))
	  (if (&& (== (head string) #\,)
		  (< 1 (length string)))     ;; otherwise it might be tuple (, foo)
	      (if (== (!! string 1) #\@)
		  ;; Is a spliced list
		  (sPList '(`++
			    (bq-transform-lst ([]) curlst)
			    (sPList '(`++
				      (PSym loc (tail (tail string)) qual)
				      (bq-transform-lst xs ([]))))))
		  ;; Eval symbol
		  (bq-transform-lst xs
				    (: (PList loc '(`toParseTree
						    (PSym loc (tail string) qual))
					      noPP)
				       curlst)))
	      ;; Normal symbol. Quote it!
	      (bq-transform-lst xs
				(: (PList loc ([] `PSym
						  `noSrcSpan
						  (PString loc string)
						  (PString loc qual))
					  noPP)
				   curlst))))
	 ;; Escaped lists of the form ,(foobar a b)
	 ((PList loc lst (, pre post))
	  (if (&& (< 0 (length pre)) 
		  (== (head pre) #\,))
	      (bq-transform-lst xs
				(: (PList loc ([] `toParseTree
						  (PList loc lst noPP))
					  noPP)
				   curlst))
	      (bq-transform-lst xs
				(: (sPList '(`PList
					     `noSrcSpan
					     (bq-transform-lst lst ([]))
					     `noPP)) 
				   curlst))))
	 ;; Autoquoting of strings
	 ;; The problem with strings are that they are actually a type synonym for [Char] 
         ;; As type synonyms can be part of instance declaration but toParseTree
	 ;; relays on instance decls., we have to manually convert strings here
	 ((@ s (PString loc str))
	  (bq-transform-lst xs
			    (: (PList loc
				      '(`PString
					`noSrcSpan
					(PString noSrcSpan str))
				      noPP)
			       curlst)))
	 (other (bq-transform-lst xs
				  (: (PList (ploc other) '(`toParseTree
							   other)
					    noPP)
				     curlst)))))))

(define (backquote-dispatcher-on-lists ks kf pt)
    (case pt
      ((PList loc lst (, "`" ""))
       (ks (sPList '(`PList
		    `noSrcSpan
		    (bq-transform-lst lst ([]))
		    `noPP))))
      (_ (kf pt))))

(define (backquote-dispatcher ks kf)
    (backquote-dispatcher-on-syms ks (backquote-dispatcher-on-lists ks kf)))

