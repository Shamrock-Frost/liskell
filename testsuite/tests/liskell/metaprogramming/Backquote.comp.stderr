
Backquote.lsk:17:58:
    Couldn't match expected type `TransformationMonad (HsExpr.LHsExpr RdrName.RdrName)'
	   against inferred type `(ParseTree -> a) -> (ParseTree -> a) -> a'
    In the second argument of `backquote-dispatcher-on-syms', namely
	`e'
    In the first argument of `LskEnv', namely
	`backquote-dispatcher-on-syms trf_expr e'
    In the first argument of `return', namely
	`LskEnv (backquote-dispatcher-on-syms trf_expr e) p t d'
Loading package base ... linking ... done.
Loading package template-haskell ... linking ... done.
Loading package readline-1.0 ... linking ... done.
Loading package unix-1.0 ... linking ... done.
Loading package Cabal-1.1.7 ... linking ... done.
Loading package regex-base-0.71 ... linking ... done.
Loading package regex-posix-0.71 ... linking ... done.
Loading package regex-compat-0.71 ... linking ... done.
Loading package haskell98 ... linking ... done.
Loading package ghc-6.7 ... linking ... done.
